pyqtorch/time_dependent/integrators/adaptive.py

```diff
@@ -0,0 +1,190 @@
+            if t + dt >= t1:
+                cache = (dt, error)
+                dt = t1 - t
+
+            # compute the next step
+            ft_new, y_new, y_err = self.step(t, y, ft, dt, self.odefun)
+
+            error = self.get_error(y_err, y, y_new)
+            if error <= 1:
+                t, y, ft = t + dt, y_new, ft_new
+
+            # check max steps are not reached
+            self.increment_step_counter(t)
+
+        dt, error = cache
+        return y, ft, dt, error
+
+    def increment_step_counter(self, t: float) -> None:
+        """Increment the step counter and check for max steps."""
+        self.step_counter += 1
+        if self.step_counter == self.options.max_steps:
+            raise RuntimeError(
+                "Maximum number of time steps reached in adaptive time step ODE"
+                f" solver at time t={t:.2g} (`max_steps={self.options.max_steps}`)."
+                " This is likely due to a diverging solution. Try increasing the"
+                " maximum number of steps, or use a different solver."
+            )
+
+    @torch.no_grad()
+    def get_error(self, y_err: Tensor, y0: Tensor, y1: Tensor) -> float:
+        """Compute the error of a given solution.
+
+        See Equation (4.11) of [1].
+        """
+        scale = self.options.atol + self.options.rtol * torch.max(y0.abs(), y1.abs())
+        return float(hairer_norm(y_err / scale).max().item())
+
+    @torch.no_grad()
+    def init_tstep(
+        self, t0: float, y0: Tensor, f0: Tensor, fun: Callable[[float, Tensor], Tensor]
+    ) -> float:
+        """Initialize the time step of an adaptive step size integrator.
+
+        See Equation (4.14) of [1] for the detailed steps. For this function, we keep
+        the same notations as in the book.
+        """
+
+        sc = self.options.atol + torch.abs(y0) * self.options.rtol
+        d0, d1 = (
+            hairer_norm(y0 / sc).max().item(),
+            hairer_norm(f0.to_dense() / sc).max().item(),
+        )
+
+        if d0 < 1e-5 or d1 < 1e-5:
+            h0 = 1e-6
+        else:
+            h0 = 0.01 * d0 / d1
+
+        y1 = y0 + h0 * f0
+        f1 = fun(t0 + h0, y1)
+        d2 = hairer_norm((f1 - f0).to_dense() / sc).max().item() / h0
+        if d1 <= 1e-15 and d2 <= 1e-15:
+            h1 = max(1e-6, h0 * 1e-3)
+        else:
+            h1 = (0.01 / max(d1, d2)) ** (1.0 / float(self.order + 1))
+
+        return min(100 * h0, h1)
+
+    @torch.no_grad()
+    def update_tstep(self, dt: float, error: float) -> float:
+        """Update the time step of an adaptive step size integrator.
+
+        See Equation (4.12) and (4.13) of [1] for the detailed steps.
+        """
+        if error == 0:  # no error -> maximally increase the time step
+            return dt * self.options.max_factor
+
+        elif error <= 1:  # time step accepted -> take next time step at least as large
+            return float(
+                dt
+                * max(
```

Comment:
Weird indentation.

Reply:
That's the outcome of `black` formatter.